Expose line/column/character span position information from parsed EDN streams

victorporof opened the pull request:
Fixes https://github.com/mozilla/mentat/issues/258

Signed-off-by: Victor Porof <victor.porof@gmail.com>

ncalexan commented on edn/src/edn.rustpeg:
```diff
@@ -29,31 +30,60 @@ use types::Value;
 // TODO: Support discard
 
 pub nil -> Value =
-    "nil" { Value::Nil }
```
I feel like `rust-peg` probably solves this problem generically, marking up your tokens with spans automatically.  Did you investigate this?

victorporof replied:
Need to add a #position myself. This is WIP.

ncalexan commented on edn/src/types.rs:
```diff
@@ -17,9 +17,21 @@ use symbols;
+
+/// Span represents the current offset into the input string.
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct Span(usize);
```
Let's do what the Rust compiler and other places do, which is that a span is a start and an end (or a length, but I prefer end) into the input buffer.

victorporof replied:
Rustpeg only gives positions. Compute the length myself?

victorporof commented:
Fine -_-

ncalexan commented on edn/src/types.rs:
```diff
@@ -214,12 +264,6 @@ impl Value {
     }
 }
 
-impl From<f64> for Value {
```
One type can have many `From` implementations.  If anything, I want `edn::Value` to have `From` for _more_ types (when there's no ambiguity).  Can you explain why you removed this?

ncalexan replied:
Oh, I see, you made these macros.  For just a few implementations I think this trickery is overkill, but it's your call.

victorporof replied:
It's a leftover from the previous implementation, but makes this a bit more symmetrical/generic so I'll keep it.

ncalexan commented on edn/src/types.rs:
```diff
@@ -248,6 +292,18 @@ impl Ord for Value {
     }
 }
 
+impl PartialOrd for ValueAndSpan {
```
This surprises me.  I don't expect `ValueAndSpan` instances with _different_ spans to compare or order equal.

@rnewman, do you?

victorporof replied:
Not strictly necessary, I'll remove it if I have to.

ncalexan commented on edn/src/types.rs:
```diff
@@ -321,44 +377,45 @@ mod test {
 
     #[test]
     fn test_print_edn() {
         let string = "[ 1 2 ( 3.14 ) #{ 4N } { foo/bar 42 :baz/boz 43 } [ ] :five :six/seven eight nine/ten true false nil #f NaN #f -Infinity #f +Infinity ]";
```
You're mapping `Value` into `ValueAndSpan`, manually.  Why not map `ValueAndSpan` onto `Value`, which we need for all the other consumers?  As written, this isn't testing the span business anyway.

victorporof replied:
If I duplicate the hierarchy as described above, then yes, mapping ValueAndSpan onto Value would be possible here. Otherwise not (because Value::Collection was over ValueAndSpan)

ncalexan commented on edn/src/types.rs:
```diff
@@ -31,21 +43,33 @@ pub enum Value {
     // LinkedList is faster for appending (which we do a lot of).
     // See https://github.com/mozilla/mentat/issues/231
-    List(LinkedList<Value>),
+    List(LinkedList<ValueAndSpan>),
```
I think this should remain `Value`, for current consumers.

Then, we want a parallel hierarchy like you've defined here, which could have two types (`ValueAndSpan(inner: InnerValueAndSpan , span: Span)` like you have, and `InnerValueAndSpan` where the container values are `ValueAndSpan`, just like here.

The map between is clear.

victorporof replied:
I see. That's a lot of code duplication. No way around it?

ncalexan commented on edn/src/edn.rustpeg:
```diff
@@ -28,32 +29,61 @@ use types::Value;
+    start:#position "nil" end:#position {
+        ValueAndSpan {
+            inner: Nil,
+            span: Some(Span(start, end))
```
Ah, I see.  Nifty!

victorporof commented:
Addressed comments and rebased.

victorporof commented:
@ncalexan this is ready for review.

ncalexan commented on edn/src/types.rs:
```diff
@@ -214,9 +318,9 @@ impl Value {
-    fn from(src: f64) -> Value {
-        Value::Float(OrderedFloat::from(src))
+impl ValueAndSpan {
+    pub fn reduce(self) -> Value {
```
This is a reduction, but Rust tends to name general reductions `fold` and special cases like this `into_inner`, to be clear that you're consuming `self`.

victorporof replied:
Changed.

ncalexan replied:
You called this `fold`, right?  This isn't a general `fold`, it's `into_inner`.

ncalexan commented on edn/src/types.rs:
```diff
@@ -44,12 +44,96 @@ pub enum Value {
+    }
+}
+
+impl From<Value> for ValueAndSpan {
```
Hmm.  Not sure I want this to be trivial.

victorporof replied:
Removed.

victorporof replied:
The reason for this was because it seemed useful to have a conversion between SpannedValue and Value, but in order to have that, we'd also need to have something from ValueAndSpan to value, because SpannedValue collections have ValueAndSpan children. Not having trivial conversion here meant that a lot of code (like all the from_foo, from_keyword, from_namespace etc. methods would've had to be duplicated. I've ultimately done it with macros to avoid any trivial conversion from Value to ValueAndSpan.

ncalexan commented:
Oh -- and do you think there might be a reason to just accept a generic type `S` for the span?  I could imagine other situations where one might want to mark up trees (with weights, scores, metadata) and the structure is there.

victorporof commented:
> Oh -- and do you think there might be a reason to just accept a generic type S for the span? I could imagine other situations where one might want to mark up trees (with weights, scores, metadata) and the structure is there.

I don't think there's any value in having that at this time.

victorporof commented:
@ncalexan this is ready for final review. I've addressed all comments.

ncalexan commented on edn/src/edn.rustpeg:
```diff
@@ -112,48 +156,69 @@ symbol_name = ( symbol_char_initial+ / "." ) ( symbol_char_subsequent* / "." )
 // It's important that float comes before integer or the parser assumes that
 // floats are integers and fails to parse
-pub value -> Value =
+pub value -> ValueAndSpan =
```
Is it not possible to do all of the `#position` stuff here, in `value`?  That is, capture the posititions before and after the `__` markers below, parse `Value` in the internal bits, and then convert to `ValueAndSpan`?  Maybe not, because of the internal whitespace that gets ignored?

Ah, as I think more about this, I think not: the recursive nature of `ValueAndSpan` denies that possibility, since the children of a container need to have the right type.  So sad.  (This might be possible with a `ValueAndSpan<S>` parameterized by an optional span type.)

victorporof replied:
As you've figured out, not possible due to the fact that SpannedValue has ValueAndSpan children in collections.

ncalexan commented on edn/src/types.rs:
```diff
@@ -44,12 +43,63 @@ pub enum Value {
 }
 
-use self::Value::*;
+/// SpannedValue is the same as Value but used in ValueAndSpan
```
nit: `SpannedValue` is parallel to `Value` but container types have `ValueAndSpan` children.

victorporof replied:
I'll update the comment. Should be obvious though.

ncalexan commented on edn/src/types.rs:
```diff
@@ -44,12 +43,63 @@ pub enum Value {
     // TODO: Make sure float syntax is correct, handle NaN and escaping.
     // See https://github.com/mozilla/mentat/issues/232
     fn fmt(&self, f: &mut Formatter) -> ::std::fmt::Result {
+        use self::Value::*;
```
Nifty -- I did not know this was possible in a function scope.  TIL!

ncalexan commented on edn/src/types.rs:
```diff
@@ -115,9 +187,9 @@ macro_rules! def_is {
-/// in an Option, like `<Option<i64>`.
+/// Creates `as_$TYPE` helper functions for Value or SpannedValue, like
+/// `as_integer()`, which returns the underlying value representing the
+/// original variable wrapped in an Option, like `<Option<i64>`.
```
nit: here and below, drop the outer `<>`.  So "like `Option<i64>`".

victorporof replied:
Ok.

ncalexan commented on edn/tests/tests.rs:
```diff
@@ -15,207 +15,424 @@ extern crate ordered_float;
 
 #[test]
 fn test_nil() {
-    assert_eq!(nil("nil").unwrap(), Nil);
```
Throughout this whole file, you've converted a large number of tests from comparing against `T` to comparing against the `edn::Value` version of `T`.  That's a valid choice, but why did you choose it?  And could you have done this as a mechanical `Pre:` patch?  (With a commit comment explaining why you wanted this or what it makes easier later, etc, etc.)  As it stands I can't see what's really new here and what's just different.

victorporof replied:
To avoid even more changes to this file.

victorporof replied:
I think I might be able to get away with a bit less changes, but the crux of the issue is that both SpannedValue and Value are imported the the tests file, and the edn parser returns ValueAndSpan now. Without those changes, we would't be checking Values anymore, but SpannedValues, so some kind of conversion needs to be done: this is via folding, or maintaining the previous intent to test Value, or something else. My approach was one of many possible ones. To make life easier for other people, I chose to drop the `*` import and make things specific.

ncalexan commented on tx-parser/tests/parser.rs:
```diff
@@ -26,7 +26,7 @@ fn test_entities() {
 [:db/retract 102 :test/b "w"]]"#;
 
-    let edn = parse::value(input).unwrap();
+    let edn = parse::value(input).unwrap().fold();
```
I know I suggested `into_inner`, but maybe we should be explicit and say `without_spans`?  That'll certainly read easiest...

victorporof replied:
Will do.

ncalexan commented on edn/src/types.rs:
```diff
@@ -248,7 +389,32 @@ impl Ord for Value {
+
+impl Ord for SpannedValue {
+    fn cmp(&self, other: &SpannedValue) -> Ordering {
+        Value::from(self.clone()).cmp(&Value::from(other.clone()))
```
This is going to be very expensive.  Why do either of these types implement `Ord` and `PartialOrd` at all?  What is the desired ordering?  Is it a domain concept?  Is it because we're using `edn::Value` in some ordered data structure?  Maybe the inner `BTreeSet` and `BTreeMap`?  We might not want to do that.

Can we be clear on what we're doing here and why?

victorporof replied:
As discussed on Slack, cloning isn't necessary, just a shortcut I took in order to get something that is landable (as in: provides value that doesn't exist yet, is better than nothing etc.). Changing this implementation to avoid cloning is relatively trivial, sufficient for a good followup good first bug, but I'll do it in this PR.

victorporof commented:
@ncalexan Addressed comments, avoided cloning in `Ord` implementation, and reduced test code churn as much as I could (although this seems to be strictly worse in terms of readability, because Value and Spanned value both have the same types so it's a bit hard to determine what exactly is being tested. Ready for another review.

victorporof commented:
@ncalexan I've lost some changes in a rebase, stand by.

victorporof commented:
Reflog rescued me, ready for review.

victorporof commented:
@ncalexan gentle review ping.

victorporof commented:
@ncalexan The failure is due to something landing in the meantime. Easy fix.

victorporof commented:
@ncalexan Fixed.
